### PR TITLE
dbt project v2

### LINKED ISSUE #2300: dbt_project.yml v2
### Describe the feature
dbt_project.yml should get a version 2. We can denote it by adding an optional `config-version:` field that defaults to v1 and encourage users to migrate to v2 by creating a migration script.

Many of these changes will enable/make it easier to implement #2269 and #2283 - in fact, v2 will probably be required for those features. It'll also make fixing #2265 a lot easier.

The changes:
 - The `models` config block will be split out into `seeds`, `models`, `analyses`, `sources`, etc.
 - In those config blocks, every entry will be used as a part of the FQN hierarchy, except for the key named `config` - that key will denote the actual values to be applied to the config.
   - This will resolve some ambiguities
   - Directories and models named `config` in the search paths will be an error.
 - Give `vars` its own top-level key. `vars` will be project-scoped only, no full paths/FQNs
   - this will still make it possible to override dependency vars, which is important
   - but we won't have to do a bunch of hierarchy-parsing from configs.
   - `vars` in config blocks will be treated as a config value named `vars`.

We should provide a migration path: 0.17.0 will support both versions (but you can't have the "good stuff" on v1) and emit deprecation warnings. 0.18.0 will only support v2.

### Describe alternatives you've considered
We can just not do it.

### Additional context
N/A

### Who will this benefit?
anyone developing sources

### PR DESCRIPTION
**Pull request by beckjake**
resolves #2300 

### Description
This PR adds support for a dbt_project.yml v2, as described in #2300

Details:
 - The single source of truth for values in the config is now the `NodeConfig` and its descendants. The v1 config still uses a special class that implements the legacy behavior, for backwards compatibility reasons.
 - Cleaned up snapshot config definitions to be mutually exclusive and serialize/deserialize more easily and consistently.
 - If a node's project and the active project are different config versions, the higher version one is "downgraded" into v1
 - Adapter specific configs are now typed JsonSchemaMixins
   - Removed the deprecated bigquery functionality of partition_by supporting strings
 - RuntimeConfig is now explicitly aware of its dependency packages, removing the need to pass the collection of packages around
 - I moved a number of things around
 - Added type annotations for many modules that previously didn't have them
 - Sources now have Configs
 	- You can set the `enabled` flag
 - Added a number of v1/v2 compatibility shims that we can remove when this is done
 - Split configuration file rendering into components 

### Checklist
 - [X] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
 - [X] I have run this code in development and it appears to resolve the stated issue
 - [X] This PR includes tests, or tests are not required/relevant for this PR
 - [ ] I have updated the `CHANGELOG.md` and added information about my change to the "dbt next" section.

### CODE REVIEW

**Comment by beckjake** [comment]
I've rebased this again to remove the merge conflict and (I hope) fix the last of those minor test failures in the process.

**Comment by drewbanin** [comment]
Things to check:
 - warning for configs that are useless
 - update description to not show bit about quoting configs (this will help me later @ documentation time) 
 - replace `config:` block with `+` syntax. Make this optional, but always allowed for config declarations

**Comment by beckjake** [comment]
> warning for configs that are useless

Locally this seems to work ok for me now, but I did mess with that code in this most recent change!

**Comment by beckjake** [comment]
Ok, I've switched out `config` for an optional `+` prefix - we test for it in the column types test, so I feel pretty good about it.

**Comment by drewbanin** [comment]
Ok - I played around with this some more. Can we:
 - Make `sources:` participate in useless config checking? It looks like useless source configs are not currently checked / warned
 - Disallow specifying `vars:` in the `dbt_project.yml` v2 spec `models:` section?
   - This is going to be the most likely failure mode IMO -- if a user specifies that they're using config version 2, then we should at least _warn_ about `vars:` declarations as they will be unused

I'm going to start digging into the code in this diff. Will follow up with other thoughts inline while reviewing / testing. Looking good so far though!

File: core/dbt/config/runtime.py [hunk 1]
```diff
@@ -174,6 +221,155 @@ def get_metadata(self) -> ManifestMetadata:
...
+            return self._get_v2_config_paths(config, path, paths)
+        else:
+            return self._get_v1_config_paths(config, path, paths)
+
+    def get_resource_config_paths(self) -> Dict[str, PathSet]:
+        """Return a dictionary with 'seeds' and 'models' keys whose values are
+        lists of lists of strings, where each inner list of strings represents
+        a configured path in the resource.
+        """
+        return {
```
Comment (drewbanin) [inline core/dbt/config/runtime.py:277]: Can we add sources in here too?

File: core/dbt/context/context_config.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+    def _update_from_config(
+        self, result: T, partial: Dict[str, Any], validate: bool = False
+    ) -> T:
+        return result.update_from(
+            partial,
+            self.active_project.credentials.type,
+            validate=validate
+        )
+
+    # TODO: is fqn[0] always the project name?
```
Comment (drewbanin) [inline core/dbt/context/context_config.py:140]: 👁️ 👁️

File: core/dbt/deprecations.py [hunk 3]
```diff
@@ -92,23 +92,16 @@ class ModelsKeyNonModelDeprecation(DBTDeprecation):
...
-
-    - Provided partition_by: {raw_partition_by}
-
-
-    - dbt inferred: {inferred_partition_by}
-
+class DbtProjectYamlDeprecation(DBTDeprecation):
+    _name = 'dbt-project-yaml-v1'
+    _description = '''\
+    The existing dbt_project.yml format has been deprecated. dbt_project.yml
```
Comment (drewbanin) [inline core/dbt/deprecations.py:98]: I can add a link + update copy here

File: plugins/bigquery/dbt/adapters/bigquery/impl.py [hunk 4]
```diff
@@ -126,6 +77,16 @@ def _stub_relation(*args, **kwargs):
     )
 
 
+@dataclass
+class BigqueryConfig(AdapterConfig):
+    cluster_by: Optional[Union[List[str], str]] = None
+    partition_by: Optional[Dict[str, Any]] = None
+    kms_key_name: Optional[str] = None
+    labels: Optional[Dict[str, str]] = None
+    # TODO: should this accept `str` and `int`, too?
```
Comment (drewbanin) [inline plugins/bigquery/dbt/adapters/bigquery/impl.py:86]: Is this for `partitions`? I think a list of ints would also be valid for some use cases, but users can always pipe a list of ints to `| map('string')` or similar if needed

**Comment by drewbanin** [comment]
Some things I've noticed while testing:

-----------------

Can we specify which packages are configured with v1 `dbt_project.yml` files? I was confused by the warning I was seeing locally until I realized that deprecation warning was coming from a dependency package.

------------------

It looks like the `vars` dict we're supplying to nodes includes the top-level package name keys. I see this error when misconfigured a model's vars:

```
Compilation Error in model vars (models/vars/enabled.sql)
  Required var 'enabled' not found in config:
  Vars supplied to enabled = {
      "debug": {
          "key": "ok",
          "mydict": {
              "a": 1
          }
      },
      "key": "ok",
      "mydict": {
          "a": 1
      }
  }

  > in model vars (models/vars/enabled.sql)
  > called by model vars (models/vars/enabled.sql)
```

My corresponding `vars:` config looks like:

```
vars:
    debug:
        key: ok
        mydict: {a: 1}
```

Should `debug` (the package name) be a member in the `var()` dict? 

--------

We should add more error handling around var values. The following `vars:` values make dbt fail hard:
```
vars:

vars:
    debug: something

vars:
    debug:
```

-------

If we can figure out a way to warn for `vars:` found in a v1 `dbt_project.yml` file, I think it would make it much easier for folks to upgrade to the new spec!

File: core/dbt/context/context_config.py (hunk 2 above)
Comment (beckjake) [reply core/dbt/context/context_config.py:140]: Oh yeah... the answer is "yes, but I don't trust that to always be true"

File: plugins/bigquery/dbt/adapters/bigquery/impl.py (hunk 4 above)
Comment (beckjake) [reply plugins/bigquery/dbt/adapters/bigquery/impl.py:86]: Yeah. That sounds reasonable to me - I think it previously did, but I don't think it necessarily was intended to do so.

**Comment by beckjake** [comment]
>Should debug (the package name) be a member in the var() dict?

I came around (for compatibility reasons) to the idea of `vars` having two tiers - global and package-scoped. I don't know if it's the best idea, but it seemed to make it easier to migrate. So yes, it could be!

**Comment by drewbanin** [comment]
> I came around (for compatibility reasons) to the idea of vars having two tiers - global and package-scoped.

Yep, that makes sense. I didn't consider that case when I last looked at this last, but I buy that global and package scoped tiers should be separately addressable. This does get wonky if a var in a package masks a package name, eg:

```
vars:
  my_package:
    other_package: something
  other_package:
    myvar: true
```

Here, the `my_package.other_package` var is going to mask the vars defined for `other_package`. I'm ok with that, but it would be even better if we added a warning :)

**Comment by beckjake** [comment]
In your example, a model in `my_package` that looked for `var('other_package')` would get back `'something'`. A model outside `my_package` that looked for `var('other_package')` would get back `{'myvar': True}`.

I'm not sure if that's what I'd expect, but I don't think I'd be surprised?

File: core/dbt/deprecations.py (hunk 3 above)
Comment (drewbanin) [reply core/dbt/deprecations.py:98]: Let's go for something like:

```
dbt v0.17.0 introduces a new config format for the dbt_project.yml file. Support
for the existing version 1 format will be removed in a future release of dbt.
The following packages are currently configured with config version 1:
 - debug
 - some_package

For upgrading instructions, consult the documentation:
  https://docs.getdbt.com/docs/guides/migration-guide/upgrading-to-0-17-0
```

File: core/dbt/parser/manifest.py [hunk 5]
```diff
@@ -331,7 +331,16 @@ def load_all(
         macro_hook: Callable[[Manifest], Any],
     ) -> Manifest:
         with PARSING_STATE:
-            projects = load_all_projects(root_config)
+            projects = root_config.load_dependencies()
+            v1_configs = []
```
Comment (drewbanin) [inline core/dbt/parser/manifest.py:335]: I'm testing this out locally, and I'm (strangely) only seeing one project or package name in the deprecation warning. My setup:
 - has a v1 `dbt_project.yml` file in my root project
 - has a v1 `dbt_project.yml` file in a dependency package

Any idea what might be going wrong?

File: core/dbt/config/renderer.py [hunk 6]
```diff
@@ -1,137 +1,212 @@
...
+            return True
 
         first = keypath[0]
         # run hooks
         if first in {'on-run-start', 'on-run-end', 'query-comment'}:
-            return True
+            return False
         # models have two things to avoid
-        if first in {'seeds', 'models', 'snapshots'}:
+        if first in {'seeds', 'models', 'snapshots', 'seeds'}:
```
Comment (swiffer) [inline core/dbt/config/renderer.py:75]: @beckjake it's already in there (and line 98)

**Pull request merged**